packages/react/src/components/NumberInput/NumberInput.js

```diff
@@ -310,12 +318,25 @@ class NumberInput extends Component {
     const inputWrapperProps = {};
     let errorId = null;
     let error = null;
-    if (
-      invalid ||
-      (!allowEmpty && this.state.value === '') ||
-      this.state.value > max ||
-      this.state.value < min
-    ) {
+
+    let isInputInvalid;
+
+    // If the user supplied `invalid` through props, we'll defer to the passed in value
+    if (invalid) {
+      isInputInvalid = true;
+    } else {
+      // Otherwise, if we don't allow an empty value then we check to see
+      // if the value is empty, or if it is out of range
+      if (!allowEmpty && this.state.value === '') {
+        isInputInvalid = true;
+      } else {
+        if (this.state.value > max || this.state.value < min) {
```

Comment:
```suggestion
        if (this.state.value !== '' && (this.state.value > max || this.state.value < min)) {
```

Looks that #4997 was not fixed in this PR's code. #4997 specifically asks for the following scenario:

> ... meaning that an empty string is only consider valid (despite `allowEmpty`) if **the range includes zero**

The problem is observed by setting `10` to `min`, `100` to `max`, and `true` to `allowEmpty`. Above suggested change seems to fix the problem.

Reply:
Ah yeah, good catch and thanks for looking into this!